File: apps/mobile-android/app/src/main/java/com/univalle/pedrochacolla/ui/dashboard/MapNightMistEffect.kt
package com.univalle.pedrochacolla.ui.dashboard

import android.graphics.Canvas
import android.graphics.Paint
import android.graphics.Path
import android.graphics.RadialGradient
import android.graphics.Shader
import com.univalle.pedrochacolla.utils.map.AmbientTickOptions
import com.univalle.pedrochacolla.utils.map.MapAmbientZone
import kotlin.math.cos
import kotlin.math.sin
import kotlin.random.Random

class MapNightMistEffect {
    var intensity = 0.35f

    private data class Veil(var bx: Float, var by: Float, val r: Float, var phase: Float)

    private val veils = mutableListOf<Veil>()
    private var containsPoint: ((Float, Float) -> Boolean)? = null
    private var globalPhase = 0f
    private val fillPaint = Paint(Paint.ANTI_ALIAS_FLAG)

    fun clear() { veils.clear(); globalPhase = 0f }
    fun setContainsPoint(fn: ((Float, Float) -> Boolean)?) { containsPoint = fn }

    private fun inZone(bx: Float, by: Float) = containsPoint?.invoke(bx, by) ?: true

    fun tick(options: AmbientTickOptions, dt: Float = 1f) {
        options.containsPoint?.let { containsPoint = it }
        val b = options.bounds
        if (b.maxX - b.minX <= 0f || b.maxY - b.minY <= 0f) return
        val inten = maxOf(0.15f, intensity)
        val target = MapAmbientZone.parkParticleTarget(2, 6, inten)
        while (veils.size < target) {
            val spanX = b.maxX - b.minX
            val spanY = b.maxY - b.minY
            var bx = b.minX + Random.nextFloat() * spanX
            var by = b.minY + Random.nextFloat() * spanY
            repeat(16) {
                bx = b.minX + Random.nextFloat() * spanX
                by = b.minY + Random.nextFloat() * spanY
                if (inZone(bx, by)) return@repeat
            }
            veils.add(Veil(bx, by, MapAmbientZone.parkPlanSize(22f + Random.nextFloat() * 38f), Random.nextFloat() * 6.28f))
        }
        while (veils.size > target) veils.removeAt(veils.lastIndex)
        globalPhase += 0.005f * dt
        for (v in veils) {
            v.phase += 0.008f * dt
            v.bx += sin(v.phase + globalPhase) * 0.05f * dt
            v.by += cos(v.phase * 0.85f) * 0.03f * dt
        }
    }

    fun draw(
        canvas: Canvas,
        clipPath: Path?,
        toScreen: (Float, Float) -> Pair<Float, Float>,
        screenScale: Float,
        isDarkTheme: Boolean,
        viewportW: Float,
        viewportH: Float,
    ) {
        if (!isDarkTheme) return
        val inten = maxOf(0.2f, intensity)
        val sr = MapAmbientZone.ambientScreenScale(screenScale, 1f)
        canvas.save()
        clipPath?.let { canvas.clipPath(it) }
        for (v in veils) {
            if (!inZone(v.bx, v.by)) continue
            val (x, y) = toScreen(v.bx, v.by)
            val pulse = 0.88f + sin(v.phase) * 0.12f
            val radius = v.r * sr * pulse
            fillPaint.shader = RadialGradient(
                x, y, radius,
                intArrayOf(
                    android.graphics.Color.argb((inten * 51).toInt(), 80, 120, 200),
                    android.graphics.Color.TRANSPARENT,
                ),
                floatArrayOf(0f, 1f),
                Shader.TileMode.CLAMP,
            )
            canvas.drawOval(x - radius, y - radius * 0.75f, x + radius, y + radius * 0.75f, fillPaint)
        }
        fillPaint.shader = null
        fillPaint.color = android.graphics.Color.argb((inten * 15).toInt(), 25, 40, 90)
        canvas.drawRect(0f, 0f, viewportW, viewportH, fillPaint)
        canvas.restore()
    }
}

File: apps/web-admin/src/app/features/map/utils/map-night-mist-effect.ts
import type { AmbientTickOptions } from './map-ambient-zone';
import { ambientScreenScale } from './map-ambient-zone';
import { parkPlanSize, parkParticleTarget } from './map-park-visual-scale';

interface MistVeil {
  bx: number;
  by: number;
  r: number;
  phase: number;
}

/** Bruma nocturna con tinte azul — pensada para tema oscuro del mapa. */
export class MapNightMistEffect {
  private veils: MistVeil[] = [];
  private intensity = 0.35;
  private containsPoint: ((bx: number, by: number) => boolean) | null = null;
  private globalPhase = 0;

  setIntensity(value: number): void {
    this.intensity = Math.min(1, Math.max(0, value));
  }

  setContainsPoint(fn: ((bx: number, by: number) => boolean) | null): void {
    this.containsPoint = fn;
  }

  clear(): void {
    this.veils = [];
    this.globalPhase = 0;
  }

  private inZone(bx: number, by: number): boolean {
    return !this.containsPoint || this.containsPoint(bx, by);
  }

  tick(options: AmbientTickOptions, dt = 1): void {
    if (options.containsPoint) this.containsPoint = options.containsPoint;
    const bounds = options.bounds;
    const spanX = bounds.maxX - bounds.minX;
    const spanY = bounds.maxY - bounds.minY;
    if (spanX <= 0 || spanY <= 0) return;

    const inten = Math.max(0.15, this.intensity);
    const target = parkParticleTarget(2, 6, inten);

    while (this.veils.length < target) {
      let bx = bounds.minX + Math.random() * spanX;
      let by = bounds.minY + Math.random() * spanY;
      for (let t = 0; t < 16; t++) {
        bx = bounds.minX + Math.random() * spanX;
        by = bounds.minY + Math.random() * spanY;
        if (this.inZone(bx, by)) break;
      }
      this.veils.push({
        bx,
        by,
        r: parkPlanSize(22 + Math.random() * 38),
        phase: Math.random() * Math.PI * 2,
      });
    }
    while (this.veils.length > target) this.veils.pop();

    this.globalPhase += 0.005 * dt;
    for (const v of this.veils) {
      v.phase += 0.008 * dt;
      v.bx += Math.sin(v.phase + this.globalPhase) * 0.05 * dt;
      v.by += Math.cos(v.phase * 0.85) * 0.03 * dt;
    }
  }

  draw(
    ctx: CanvasRenderingContext2D,
    clipPath: Path2D | null,
    toScreen: (bx: number, by: number) => { x: number; y: number },
    screenScale = 1,
    isDarkTheme: boolean,
    viewportW: number,
    viewportH: number,
  ): void {
    if (!isDarkTheme) return;

    const inten = Math.max(0.2, this.intensity);
    const sr = ambientScreenScale(screenScale, 1);
    ctx.save();
    if (clipPath) ctx.clip(clipPath);

    for (const v of this.veils) {
      if (!this.inZone(v.bx, v.by)) continue;
      const { x, y } = toScreen(v.bx, v.by);
      const pulse = 0.88 + Math.sin(v.phase) * 0.12;
      const radius = v.r * sr * pulse;
      const g = ctx.createRadialGradient(x, y, 0, x, y, radius);
      g.addColorStop(0, `rgba(80, 120, 200, ${inten * 0.2})`);
      g.addColorStop(0.5, `rgba(50, 80, 160, ${inten * 0.12})`);
      g.addColorStop(1, 'rgba(30, 50, 100, 0)');
      ctx.fillStyle = g;
      ctx.beginPath();
      ctx.ellipse(x, y, radius, radius * 0.75, 0, 0, Math.PI * 2);
      ctx.fill();
    }

    ctx.fillStyle = `rgba(25, 40, 90, ${inten * 0.06})`;
    ctx.fillRect(0, 0, viewportW, viewportH);

    ctx.restore();
  }
}
